lib/kube/proxy/portforward_test.go

```diff
@@ -213,3 +219,171 @@ type portForwarder interface {
+		portForwardRequest: portForwardRequest{
+			context:       context.Background(),
+			onPortForward: func(addr string, success bool) {},
+		},
+
+		Entry:                 logger,
+		sourceConn:            sourceConn,
+		targetConn:            targetConn,
+		streamChan:            make(chan httpstream.Stream),
+		streamPairs:           map[string]*httpStreamPair{},
+		streamCreationTimeout: 5 * time.Second,
+	}
+
+	go func() {
+		dataStream, err := sourceConn.CreateStream(http.Header{
+			PortForwardRequestIDHeader: []string{reqID},
+			PortHeader:                 []string{port},
+			StreamType:                 []string{StreamTypeError},
+		})
+		assert.NoError(t, err)
+		h.streamChan <- dataStream
+		errStream, err := sourceConn.CreateStream(http.Header{
+			PortForwardRequestIDHeader: []string{reqID},
+			PortHeader:                 []string{port},
+			StreamType:                 []string{StreamTypeData},
+		})
+		assert.NoError(t, err)
+		h.streamChan <- errStream
+		// Close the source after the streams are processed to unblock the call.
+		sourceConn.Close()
+	}()
+	// run the port forward proxy. it will read the h.streamChan and
+	// process the streams synchronously. Once the streams are processed,
+	// the sourceConn will be closed and the proxy will exit the run loop.
+	h.run()
+	// targetConn is closed once all streams are removed. It is an hack to
+	// unblock the targetConn.waitForClose() call otherwise it will block
+	// forever.
+	targetConn.waitForClose()
+	// Make sure all streams are removed from the connection.
+	require.Empty(t, sourceConn.streams)
+	// Make sure all streams are removed from the connection.
+	require.Empty(t, targetConn.streams)
+	require.True(t, sourceConn.wereStreamsClosed())
+	require.True(t, targetConn.wereStreamsClosed())
+}
+
+var _ httpstream.Stream = (*fakeSPDYStream)(nil)
+
+type fakeSPDYStream struct {
+	closed     bool
+	headers    http.Header
+	identifier uint32
+}
+
+func (f *fakeSPDYStream) Read(p []byte) (n int, err error) {
+	return 0, io.EOF
+}
+
+func (f *fakeSPDYStream) Write(p []byte) (n int, err error) {
+	return len(p), nil
+}
+
+func (f *fakeSPDYStream) Headers() http.Header {
+	return f.headers
+}
+
+func (f *fakeSPDYStream) Reset() error {
+	return nil
+}
+
+func (f *fakeSPDYStream) Identifier() uint32 {
+	return f.identifier
+}
+
+func (f *fakeSPDYStream) Close() error {
+	f.closed = true
+	return nil
+}
+
+var _ httpstream.Connection = (*fakeSPDYConnection)(nil)
+
+type fakeSPDYConnection struct {
+	count        int
+	streams      map[uint32]*fakeSPDYStream
+	streamsSlice []*fakeSPDYStream
+	closed       chan bool
+	once         sync.Once
+}
+
+func newfakeSPDYConnection() *fakeSPDYConnection {
+	return &fakeSPDYConnection{
+		streams: make(map[uint32]*fakeSPDYStream),
+		closed:  make(chan bool),
+	}
+}
+
+// CreateStream creates a new Stream with the supplied headers.
+func (f *fakeSPDYConnection) CreateStream(headers http.Header) (httpstream.Stream, error) {
```

Comment:
We don't expect concurrent access here, right?

Should we use a lock just in case this changes in the future, or do you think that's too unlikely?